### DISCUSSION TITLE
Automatic seed - key response

### DISCUSSION THREAD

**Discussion by Feezex** [jglim/CaesarSuite#64] [General]
I do like v2 branch alot, playing with some custom sequences  here.
I cant figure out how i can use unlockEcu abilities here to response automatically for a seed request.
As always just having idea ,hope you can help to implement it.
Lets say im working with ki164, since theres dozens of definitions i guerss each one must have exact name as identified by cff variant, or its  Variant ID. Its seems clear if ecu variant is a single, or algo is one for all variants. 
Possibly UnlockEcu db needs one more field to keep variant ID, if its defined Diogenes then can search for ID match for auto response.
Lets say we have 100 ecus defined , and just 5 of them have "variant ID" string, then Diogenes will check up matchings between those 5 ID's 
 Other side of this - some IC's have another type of request\response to rise security level.
Seems all family for KiAlgo1 (164 171 203 203m 211 ...) uses same sequence with 31 FB 10 request for seed.
So this one also must be defined? to catch up incoming message 71 FB + seed and answer with 31 FA 10 + key.

**Comment by jglim** [comment]
That's something that I'd like as well, however I can't think of a clean and organized way to do this properly, because there are too many edge cases to handle:

- Variant detection: the request type can vary (e.g 1A 8x, or 22 F1 xx), the response and the variant id position and length may be different across ECUs.

- Session state: same issue, modern ECUs are more straightforward, older ones behave differently

- Request types: common way to elevate access is through 27 xx, but some ECUs like the IC211/IC209 require 27 01 xx. KIAlgo1 seems to use something completely different (31 FB ..) as you explained too.

- Unlock stages: sometimes the desired level (e.g. BC) requires going through a prior level (9A).

MB solves this by using scripts that are embedded in the CBF or SMR-D and this seems like the most reasonable approach. 

Diogenes will require some changes:

- A scripting engine, likely cs-script if it happens. Older Vediamo versions uses their own interpreter which we are emulating in v2, I believe the newer ones have a java backend.
- A minimal, stable interface so that scripts don't just break when the application or its libraries are updated.
- UnlockECU should be added as a submodule or nuget package. Since v2 is now based on .net core, this should be possible now, though I am still unsure on how to correctly bundle the db.json file

Another repo is likely required for community scripts, examples, documentation. Embedding a general scripting engine has an unfortunate side effect of introducing the possibility of malicious scripts from third parties, and it is unlikely that the typical user is capable of checking if a script is safe.
